### ISSUE TITLE
`suspicious use of binary operator` wrongly flags logical operators in conditions

### ISSUE THREAD

**Issue by pwnorbitals** [rust-lang/rust-clippy#7084]
Lint name: **suspicious use of binary operator**

I tried this code (uses [impl_ops ](https://docs.rs/impl_ops/0.1.1/impl_ops/)crate):

```rust
impl_op_ex!(- |a : &Struct, b : &Struct| -> Struct{
    if a.getA() == b.getA() && a.getB() == b.getB() {
        return Struct::new(a.getData() - b.getData(), a.getA(), a.getB());
    }

    todo!("A or B change on substraction is not yet supported");
});
```

I expected to see this happen: no error or warning is thrown

Instead, this happened: error is thrown on the `&&` logical operator in the `if` expression

### Meta

- `cargo clippy -V`: clippy 0.0.212 (cb75ad5d 2021-02-10) 
- `rustc -Vv`:
  ```
  rustc 1.50.0 (cb75ad5db 2021-02-10)
  binary: rustc 
  commit-hash: cb75ad5db02783e8b0222fee363c5f63f7e2cf5b
  commit-date: 2021-02-10
  host: x86_64-pc-windows-msvc
  release: 1.50.0 
  ```

**Comment by Y-Nak** [comment]
`impl_ops` seems to use closure to implement the traits, so minimum reproducer would be
```
#![deny(clippy::suspicious_arithmetic_impl)]
use std::ops::Sub;

struct Foo {
    x: i32,
    y: i32,
}

impl Foo {
    fn new(x: i32, y: i32) -> Self {
        Self { x, y }
    }
}

impl Sub for Foo {
    type Output = Self;
    fn sub(self, rhs: Self) -> Self::Output {
        {
            (|a: &Foo, b: &Foo| -> Foo {
                {
                    if a.x == 0 && b.y == 0 {
                        Foo::new(a.x - b.y, a.y - b.y)
                    } else {
                        Foo::new(a.x - b.x, a.y - b.y)
                    }
                }
            })(&self, &rhs)
        }
    }
}
```
